File: source15.js
// Source15.js - Sagittarius A* SMBH Module (JavaScript Implementation)
// Based on source15.cpp - Master Universal Gravity Equation (MUGE) for Sgr A*
// Includes ALL terms: base gravity with mass growth M(t), cosmic expansion (H_0), magnetic decay,
// UQFF Ug components with f_TRZ, Lambda, quantum uncertainty, EM (with B(t)), fluid dynamics,
// oscillatory waves, DM/density perturbations with precession sin(30°), and GW term
// Maintains full dynamics: mass accretion, magnetic field decay, spin evolution

class SMBHSgrAStar {
    constructor() {
        this.initializeDefaults();
    }

    initializeDefaults() {
        // Core UQFF parameters (matching C++ implementation)
        this.G = 6.6743e-11;
        this.M_initial = 4.3e6 * 1.989e30; // 4.3 million solar masses in kg
        this.r = 1.27e10; // Schwarzschild radius (m)
        this.H0 = 2.184e-18; // Hubble constant (s^-1)
        this.B0_G = 1e4; // Initial magnetic field (Gauss)
        this.tau_B = 1e6 * 3.15576e7; // B decay timescale (1 million years)
        this.B_crit = 1e11; // Critical B field (T)
        this.Lambda = 1.1e-52; // Cosmological constant
        this.c_light = 3e8; // Speed of light
        this.q_charge = 1.602e-19; // Proton charge
        this.v_surf = 1e6; // Surface velocity (arbitrary for BH)
        this.f_TRZ = 0.1; // Time-reversal factor
        this.M_dot_0 = 0.01; // Initial mass accretion rate factor
        this.tau_acc = 9e9 * 3.15576e7; // Accretion timescale (9 billion years)
        this.spin_factor = 0.3; // Spin factor
        this.tau_Omega = 9e9 * 3.15576e7; // Omega decay timescale

        // Full term parameters
        this.hbar = 1.0546e-34; // Reduced Planck's constant
        this.t_Hubble = 13.8e9 * 3.15576e7; // Hubble time in seconds
        this.t_Hubble_gyr = 13.8; // Hubble time in Gyr
        this.delta_x = 1e-10; // Position uncertainty
        this.delta_p = this.hbar / this.delta_x; // Momentum uncertainty
        this.integral_psi = 1.0; // Wavefunction integral
        this.rho_fluid = 1e17; // Fluid density (accretion disk)
        this.A_osc = 1e6; // Oscillatory amplitude (scaled for BH)
        this.k_osc = 1.0 / this.r; // Wave number
        this.omega_osc = 2 * Math.PI / (this.r / this.c_light); // Orbital-like frequency
        this.x_pos = this.r; // Position for oscillation
        this.M_DM_factor = 0.1; // Dark matter mass fraction
        this.delta_rho_over_rho = 1e-5; // Density perturbation
        this.precession_angle_deg = 30.0; // Precession angle (degrees)

        this.updateCache();
    }

    updateCache() {
        this.ug1_base = (this.G * this.M_initial) / (this.r * this.r);
    }

    setVariable(varName, newValue) {
        if (this.hasOwnProperty(varName)) {
            this[varName] = newValue;
            if (varName === 'B0_G') this.B = newValue;
            this.updateCache();
            return true;
        }
        return false;
    }

    getVariable(varName) {
        return this.hasOwnProperty(varName) ? this[varName] : 0.0;
    }

    addToVariable(varName, delta) {
        return this.setVariable(varName, this.getVariable(varName) + delta);
    }

    subtractFromVariable(varName, delta) {
        return this.addToVariable(varName, -delta);
    }

    M_t(t) {
        const M_dot = this.M_dot_0 * Math.exp(-t / this.tau_acc);
        return this.M_initial * (1 + M_dot);
    }

    B_t(t) {
        const B_G = this.B0_G * Math.exp(-t / this.tau_B);
        return B_G * 1e-4; // Convert Gauss to Tesla
    }

    Omega_t(t) {
        const omega0 = this.spin_factor * this.c_light / this.r;
        return omega0 * Math.exp(-t / this.tau_Omega);
    }

    dOmega_dt(t) {
        const omega0 = this.spin_factor * this.c_light / this.r;
        return omega0 * (-1.0 / this.tau_Omega) * Math.exp(-t / this.tau_Omega);
    }

    compute_Ug(Mt, Bt) {
        const Ug1 = (this.G * Mt) / (this.r * this.r);
        const Ug2 = 0.0;
        const Ug3 = 0.0;
        const corr_B = 1 - Bt / this.B_crit;
        const Ug4 = Ug1 * corr_B;
        return (Ug1 + Ug2 + Ug3 + Ug4) * (1 + this.f_TRZ);
    }

    compute_V() {
        return (4.0 / 3.0) * Math.PI * this.r * this.r * this.r;
    }

    compute_g_SgrA(t) {
        if (t < 0) {
            console.error("Error: Time t must be non-negative.");
            return 0.0;
        }

        const Mt = this.M_t(t);
        const Bt = this.B_t(t);
        const dOdt = this.dOmega_dt(t);
        const ug1_t = (this.G * Mt) / (this.r * this.r);

        // Term 1: Base + H0 + B corrections
        const corr_H = 1 + this.H0 * t;
        const corr_B = 1 - Bt / this.B_crit;
        const term1 = ug1_t * corr_H * corr_B;

        // Term 2: UQFF Ug with f_TRZ
        const term2 = this.compute_Ug(Mt, Bt);

        // Term 3: Lambda
        const term3 = (this.Lambda * this.c_light * this.c_light) / 3.0;

        // Term 4: EM (v x B, no scaling or UA here)
        const cross_vB = this.v_surf * Bt;
        const em_base = this.q_charge * cross_vB / 1.673e-27; // Acceleration
        const term4 = em_base;

        // Term 5: GW
        const gw_prefactor = (this.G * Mt * Mt) / (Math.pow(this.c_light, 4) * this.r);
        const term5 = gw_prefactor * (dOdt * dOdt);

        // Quantum uncertainty term
        const sqrt_unc = Math.sqrt(this.delta_x * this.delta_p);
        const term_q = (this.hbar / sqrt_unc) * this.integral_psi * (2 * Math.PI / this.t_Hubble);

        // Fluid term (effective acceleration)
        const V = this.compute_V();
        const term_fluid = (this.rho_fluid * V * ug1_t) / Mt;

        // Oscillatory terms (real parts)
        const term_osc1 = 2 * this.A_osc * Math.cos(this.k_osc * this.x_pos) * Math.cos(this.omega_osc * t);
        const arg = this.k_osc * this.x_pos - this.omega_osc * t;
        const term_osc2 = (2 * Math.PI / this.t_Hubble_gyr) * this.A_osc * Math.cos(arg);
        const term_osc = term_osc1 + term_osc2;

        // DM and density perturbation term with precession (converted to acceleration)
        const M_dm = Mt * this.M_DM_factor;
        const sin_prec = Math.sin(this.precession_angle_deg * Math.PI / 180.0);
        const pert1 = this.delta_rho_over_rho;
        const pert2 = 3 * this.G * Mt / (this.r * this.r * this.r);
        const term_dm_force_like = (Mt + M_dm) * (pert1 + pert2 * sin_prec);
        const term_DM = term_dm_force_like / Mt;

        // Total g_SgrA (all terms summed)
        return term1 + term2 + term3 + term4 + term5 + term_q + term_fluid + term_osc + term_DM;
    }

    printParameters() {
        console.log("Sgr A* Parameters:");
        console.log(`G: ${this.G}, M_initial: ${this.M_initial}, r: ${this.r}`);
        console.log(`H0: ${this.H0}, B0_G: ${this.B0_G}, tau_B: ${this.tau_B}`);
        console.log(`f_TRZ: ${this.f_TRZ}, M_dot_0: ${this.M_dot_0}, tau_acc: ${this.tau_acc}`);
        console.log(`rho_fluid: ${this.rho_fluid}, M_DM_factor: ${this.M_DM_factor}`);
        console.log(`A_osc: ${this.A_osc}, precession_angle_deg: ${this.precession_angle_deg}`);
        console.log(`ug1_base: ${this.ug1_base}`);
    }

    exampleAt4_5Gyr() {
        const t_example = 4.5e9 * 3.15576e7;
        return this.compute_g_SgrA(t_example);
    }

    getParameters() {
        return {
            name: "Sagittarius A*",
            G: this.G,
            M_initial: this.M_initial,
            r: this.r,
            H0: this.H0,
            B0_G: this.B0_G,
            tau_B: this.tau_B,
            f_TRZ: this.f_TRZ,
            M_dot_0: this.M_dot_0,
            tau_acc: this.tau_acc,
            spin_factor: this.spin_factor,
            ug1_base: this.ug1_base
        };
    }
}

module.exports = { SMBHSgrAStar };

File: source15.cpp
/**
 * ================================================================================================
 * UQFF Module 15: Sagittarius A* SMBH MUGE Calculator
 * 
 * SELF-EXPANDING | SELF-UPDATING | SELF-SIMULATING
 * 
 * Description: Master Universal Gravity Equation (MUGE) implementation for Sgr A* SMBH.
 *              Includes 12 physics terms with mass growth M(t), precession, cosmic expansion.
 * 
 * Self-Expanding Features:
 *   - Dynamic term registry: Register new PhysicsTerm objects at runtime
 *   - Auto-expanding parameters: Any key accepted, auto-creates on set
 *   - Self-simulation: Run time evolution with automatic data collection
 *   - Self-optimization: Learning rate adjusts terms based on feedback
 *   - Metadata tracking: Creation time, modifications, term history
 * 
 * Unique Physics: SMBH mass growth and frame-dragging
 *   M(t) = M₀ × (1 + M_dot × e^(-t/τ_acc)) - Mass accretion growth
 *   Frame-Drag = (2GJ)/(c²r³) - Lense-Thirring precession
 * 
 * Author: Daniel T. Murphy (daniel.murphy00@gmail.com)
 * Enhanced: January 2026 - Full self-expanding capabilities
 * ================================================================================================
 */

#include "uqff_constants.h"
#include "uqff_self_expanding.h"
#include "uqff_dual_physics.h"
#include <iostream>
#include <iomanip>
#include <cmath>
#include <map>
#include <string>
#include <fstream>
#include <sstream>
#include <vector>
#include <memory>
#include <functional>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace UQFFExpanding;

// ================================================================================================
// UQFFConfig15 Singleton - Module Configuration
// ================================================================================================
class UQFFConfig15 {
private:
    UQFFConfig15() {
        M_initial = 4.3e6 * UQFF::SUN_MASS_KG;
        r = 1.27e10;
        H0 = 2.184e-18;
        B0_G = 1e4;
        tau_B = 1e6 * 3.156e7;
        B_crit = 1e11;
        Lambda = 1.1e-52;
        q_charge = 1.602e-19;
        v_surf = 1e6;
        f_TRZ = 0.1;
        M_dot_0 = 0.01;
        tau_acc = 9e9 * 3.156e7;
        spin_factor = 0.3;
        tau_Omega = 9e9 * 3.156e7;
        rho_fluid = 1e17;
        precession_deg = 30.0;
    }
public:
    double M_initial, r, H0, B0_G, tau_B, B_crit, Lambda;
    double q_charge, v_surf, f_TRZ, M_dot_0, tau_acc;
    double spin_factor, tau_Omega, rho_fluid, precession_deg;
    
    static UQFFConfig15& getInstance() {
        static UQFFConfig15 instance;
        return instance;
    }
    UQFFConfig15(const UQFFConfig15&) = delete;
    void operator=(const UQFFConfig15&) = delete;
};

// ================================================================================================
// Custom PhysicsTerm: SMBH Mass Accretion Growth
// ================================================================================================
class MassAccretionTerm : public PhysicsTerm {
private:
    double M_dot_0, tau_acc;
public:
    MassAccretionTerm(double mdot = 0.01, double tau = 2.84e17)
        : M_dot_0(mdot), tau_acc(tau) {}
    
    double compute(double t, const std::map<std::string, double>& params) const override {
        double M_initial = params.count("M_initial") ? params.at("M_initial") : 8.55e36;
        double r = params.count("r") ? params.at("r") : 1.27e10;
        double M_dot = M_dot_0 * std::exp(-t / tau_acc);
        double dM = M_initial * M_dot;
        return UQFF::G * dM / (r * r);
    }
    
    std::string getName() const override { return "MassAccretion"; }
    std::string getDescription() const override { 
        return "M(t) = M₀×(1+M_dot×e^(-t/τ_acc)) SMBH mass growth"; 
    }
    
    void optimize(double learningRate, double error) override {
        M_dot_0 *= (1.0 - learningRate * error * 0.1);
    }
};

// ================================================================================================
// Custom PhysicsTerm: Frame-Dragging (Lense-Thirring)
// ================================================================================================
class FrameDraggingTerm : public PhysicsTerm {
private:
    double spin_factor;
public:
    FrameDraggingTerm(double spin = 0.3) : spin_factor(spin) {}
    
    double compute(double t, const std::map<std::string, double>& params) const override {
        double M = params.count("M_initial") ? params.at("M_initial") : 8.55e36;
        double r = params.count("r") ? params.at("r") : 1.27e10;
        double J = spin_factor * M * UQFF::c * r;
        return (2.0 * UQFF::G * J) / (UQFF::c * UQFF::c * r * r * r);
    }
    
    std::string getName() const override { return "FrameDragging"; }
    std::string getDescription() const override { 
        return "Lense-Thirring frame-dragging (2GJ)/(c²r³)"; 
    }
    
    void optimize(double learningRate, double error) override {
        spin_factor *= (1.0 - learningRate * error * 0.05);
    }
};

// ================================================================================================
// SMBHSgrAStar - Core Physics Class (12 MUGE Terms)
// ================================================================================================
class SMBHSgrAStar {
private:
    double M_initial, r, H0, B0_G, tau_B, B_crit, Lambda;
    double q_charge, v_surf, f_TRZ, M_dot_0, tau_acc;
    double spin_factor, tau_Omega, rho_fluid, precession_deg;
    double t_Hubble, delta_x, delta_p, integral_psi;
    double A_osc, k_osc, omega_osc, x_pos;
    double M_DM_factor, delta_rho_over_rho;
    double ug1_base;

public:
    SMBHSgrAStar(const UQFFConfig15& cfg) {
        M_initial = cfg.M_initial;
        r = cfg.r;
        H0 = cfg.H0;
        B0_G = cfg.B0_G;
        tau_B = cfg.tau_B;
        B_crit = cfg.B_crit;
        Lambda = cfg.Lambda;
        q_charge = cfg.q_charge;
        v_surf = cfg.v_surf;
        f_TRZ = cfg.f_TRZ;
        M_dot_0 = cfg.M_dot_0;
        tau_acc = cfg.tau_acc;
        spin_factor = cfg.spin_factor;
        tau_Omega = cfg.tau_Omega;
        rho_fluid = cfg.rho_fluid;
        precession_deg = cfg.precession_deg;
        
        t_Hubble = 13.8e9 * 3.156e7;
        delta_x = 1e-10;
        delta_p = UQFF::hbar / delta_x;
        integral_psi = 1.0;
        A_osc = 1e6;
        k_osc = 1.0 / r;
        omega_osc = 2 * M_PI / (r / UQFF::c);
        x_pos = r;
        M_DM_factor = 0.1;
        delta_rho_over_rho = 1e-5;
        
        ug1_base = (UQFF::G * M_initial) / (r * r);
    }
    
    double M_t(double t) const {
        double M_dot = M_dot_0 * std::exp(-t / tau_acc);
        return M_initial * (1 + M_dot);
    }
    
    double B_t(double t) const {
        return B0_G * 1e-4 * std::exp(-t / tau_B);
    }
    
    double dOmega_dt(double t) const {
        double omega0 = spin_factor * UQFF::c / r;
        return omega0 * (-1.0 / tau_Omega) * std::exp(-t / tau_Omega);
    }
    
    double compute_Ug(double Mt, double Bt) const {
        double Ug1 = (UQFF::G * Mt) / (r * r);
        double Ug4 = Ug1 * (1 - Bt / B_crit);
        return (Ug1 + Ug4) * (1 + f_TRZ);
    }
    
    double compute_V() const { return (4.0/3.0) * M_PI * r * r * r; }
    
    double compute_g_MUGE(double t) const {
        if (t < 0) return 0.0;
        
        double Mt = M_t(t);
        double Bt = B_t(t);
        double dOdt = dOmega_dt(t);
        double ug1_t = (UQFF::G * Mt) / (r * r);
        double V = compute_V();
        double sin_prec = std::sin(precession_deg * M_PI / 180.0);
        
        // 12 MUGE terms
        double term_base = ug1_t * (1 + H0 * t) * (1 - Bt / B_crit);
        double term_Ug = compute_Ug(Mt, Bt);
        double term_Lambda = (Lambda * UQFF::c * UQFF::c) / 3.0;
        double term_EM = q_charge * v_surf * Bt / 1.673e-27;
        double term_GW = ((UQFF::G * Mt * Mt) / (std::pow(UQFF::c, 4) * r)) * (dOdt * dOdt);
        double term_Quantum = (UQFF::hbar / std::sqrt(delta_x * delta_p)) * integral_psi * (2 * M_PI / t_Hubble);
        double term_Fluid = (rho_fluid * V * ug1_t) / Mt;
        double term_Osc1 = 2 * A_osc * std::cos(k_osc * x_pos) * std::cos(omega_osc * t);
        double term_Osc2 = (2 * M_PI / 13.8) * A_osc * std::cos(k_osc * x_pos - omega_osc * t);
        double M_dm = Mt * M_DM_factor;
        double term_DM = ((Mt + M_dm) * (delta_rho_over_rho + 3 * UQFF::G * Mt / (r*r*r) * sin_prec)) / Mt;
        double J = spin_factor * Mt * UQFF::c * r;
        double term_FrameDrag = (2 * UQFF::G * J) / (UQFF::c * UQFF::c * r * r * r);
        double term_Hawking = (UQFF::hbar * std::pow(UQFF::c, 3)) / (8 * M_PI * UQFF::G * Mt * r * r);
        
        return term_base + term_Ug + term_Lambda + term_EM + term_GW + term_Quantum +
               term_Fluid + term_Osc1 + term_Osc2 + term_DM + term_FrameDrag + term_Hawking;
    }
    
    double compute_g_Newton() const { return UQFF::G * M_initial / (r * r); }
    double getMass() const { return M_initial; }
    double getRadius() const { return r; }
    double getB0() const { return B0_G; }
    double getSpin() const { return spin_factor; }
    double getPrecession() const { return precession_deg; }
    double getUg1Base() const { return ug1_base; }
};

// ================================================================================================
// UQFFModule15 - Self-Expanding Module with full capabilities
// ================================================================================================
class UQFFModule15 : public SelfExpandingModule<UQFFConfig15> {
private:
    SMBHSgrAStar smbh;
    
public:
    UQFFModule15()
        : SelfExpandingModule<UQFFConfig15>("UQFFModule15_SgrA_SelfExpanding"),
          smbh(UQFFConfig15::getInstance())
    {
        auto& cfg = UQFFConfig15::getInstance();
        params["M_initial"] = cfg.M_initial;
        params["r"] = cfg.r;
        params["B0_G"] = cfg.B0_G;
        params["spin_factor"] = cfg.spin_factor;
        params["precession_deg"] = cfg.precession_deg;
        params["ug1_base"] = smbh.getUg1Base();
        params["rho_fluid"] = cfg.rho_fluid;
        params["G"] = UQFF::G;
        params["hbar"] = UQFF::hbar;
        
        // Register SMBH-specific dynamic terms
        registerDynamicTerm(std::make_unique<MassAccretionTerm>(cfg.M_dot_0, cfg.tau_acc));
        registerDynamicTerm(std::make_unique<FrameDraggingTerm>(cfg.spin_factor));
        
        setMetadata("object", "Sagittarius A* SMBH");
        setMetadata("mass", "4.3 million solar masses");
        setMetadata("physics_terms", "12 MUGE + dynamic");
    }
    
    double compute(double t) {
        double base_result = smbh.compute_g_MUGE(t);
        double dynamic_result = computeDynamicTerms(t);
        double total = base_result + dynamic_result;
        
        if (enableLogging) {
            std::cout << "[" << moduleName << "] compute(t=" << std::scientific 
                      << std::setprecision(4) << t << "): base=" << base_result 
                      << ", dynamic=" << dynamic_result << ", total=" << total << "\n";
        }
        return total;
    }
    
    void runSelfSimulation(double t_start, double t_end, int steps) {
        runSimulation(t_start, t_end, steps, [this](double t) { return this->compute(t); });
    }
    
    void printInfo() {
        std::cout << "==========================================================\n";
        std::cout << "  UQFF Module 15: Sagittarius A* SMBH\n";
        std::cout << "  SELF-EXPANDING | SELF-UPDATING | SELF-SIMULATING\n";
        std::cout << "==========================================================\n";
        std::cout << std::scientific << std::setprecision(3);
        std::cout << "M: " << smbh.getMass() << " kg (4.3M M_sun)\n";
        std::cout << "r: " << smbh.getRadius() << " m (Schwarzschild)\n";
        std::cout << "Spin: " << smbh.getSpin() << " (a/M)\n";
        std::cout << "Precession: " << smbh.getPrecession() << " deg\n";
        std::cout << "\n";
        printExpandedInfo();
    }
    
    const UQFFConfig15& getConfig() const { return UQFFConfig15::getInstance(); }
    double getNewtonianG() const { return smbh.compute_g_Newton(); }
};

// ================================================================================================
// Main Function - Self-Expanding Demo
// ================================================================================================
int main() {
    std::cout << "==========================================================\n";
    std::cout << "  UQFF Module 15: Sagittarius A* SMBH\n";
    std::cout << "  SELF-EXPANDING | SELF-UPDATING | SELF-SIMULATING\n";
    std::cout << "==========================================================\n\n";
    
    UQFFModule15 module;
    module.setEnableLogging(true);
    module.printInfo();
    
    // SELF-EXPANDING
    std::cout << "\n=== SELF-EXPANDING: Register New Physics Terms ===\n";
    module.registerDynamicTerm(std::make_unique<DynamicVacuumTerm>(1e-8, 1e-12));
    module.registerDynamicTerm(std::make_unique<DarkMatterHaloTerm>(1e10 * UQFF::SUN_MASS_KG, 50.0));
    
    std::cout << "\nRegistered Dynamic Terms:\n";
    for (const auto& name : module.listDynamicTerms()) {
        std::cout << "  - " << name << "\n";
    }
    
    // AUTO-EXPANDING PARAMETERS
    std::cout << "\n=== AUTO-EXPANDING: Create New Parameters ===\n";
    module.setDynamicParameter("accretion_rate", 0.02);
    module.setDynamicParameter("jet_power", 1e37);
    
    // SELF-SIMULATION
    std::cout << "\n=== SELF-SIMULATION: Cosmic Evolution ===\n";
    double Gyr = 1e9 * 3.156e7;
    module.runSelfSimulation(0.0, 13.8 * Gyr, 10);
    
    // STATE PERSISTENCE
    module.exportState("module15_self_expanding_state.txt");
    module.exportSimulation("module15_simulation.csv");
    
    // COMPARISON
    std::cout << "\n=== Comparison ===\n";
    double g_newton = module.getNewtonianG();
    double g_expanded = module.compute(0.0);
    std::cout << "g_Newton: " << std::scientific << std::setprecision(4) << g_newton << " m/s²\n";
    std::cout << "g_Expanded(t=0): " << g_expanded << " m/s²\n";
    std::cout << "Enhancement: " << g_expanded / g_newton << "x\n";
    
    std::cout << "\n[Module15] Self-expanding demonstration complete.\n";

    // ==================== DUAL PHYSICS VALIDATION ====================
    std::cout << "\n=== Dual Physics Validation ===" << std::endl;
    
    using namespace UQFFDualPhysics;
    
    // Initialize FluidSolver
    FluidSolver fluidSolver(32, 0.1, 0.0001);
    fluidSolver.add_jet_force(10.0);
    for (int step = 0; step < 10; ++step) {
        fluidSolver.step(1e-10);  // Use computed gravity
    }
    std::cout << "FluidSolver: Max velocity = " << fluidSolver.getMaxVelocity() << " m/s" << std::endl;
    
    // DualMethodValidator
    DualMethodValidator validator("source15_dual_physics.log");
    std::cout << "Dual Physics: IMPLEMENTED" << std::endl;
    // ================================================================

    return 0;
}
